### PR TITLE
Add turbo monorepo utility

### PR DESCRIPTION
**Pull request by ErikMichelson**
### Component/Part
CI / monorepo

### Description
This PR adds [turbo](https://turbo.build/repo) as a mono-repo utility (after lerna didn't work properly).
It is currently used to build, lint and format the different parts of the project with one command from the repository root (`yarn build`, `yarn lint`, `yarn format`).
This PR changes the workflows to use one common job for linting and checking formatting of all project parts using the new turbo commands.

### Steps
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### Related Issue(s)
none

### CODE REVIEW

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

**Comment by mrdrogdrog** [comment]
To improve development and CI runs we could host this https://ducktors.github.io/turborepo-remote-cache/

**Comment by davidmehren** [comment]
We can probably also add a `test` job to turbo and then get rid of our complicated CI job setup that detects changed files, as turbo should do that for us.

File: backend/docker/Dockerfile [hunk 1]
```diff
@@ -13,7 +13,7 @@
 FROM docker.io/node:19-alpine@sha256:80844b6643f239c87fceae51e6540eeb054fc7114d979703770ec75250dcd03b as base
 # Add tini to handle signals
 # https://github.com/nodejs/docker-node/blob/main/docs/BestPractices.md#handling-kernel-signals
-RUN apk add --no-cache tini
+RUN apk add --no-cache tini libc6-compat
```
[the remark this answers is not in the document]
Comment (ErikMichelson) [reply backend/docker/Dockerfile:16]: Without this package yarn install fails completely.

[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply backend/docker/Dockerfile:16]: Yes, but neither yarn nor turbo are used in the "prod" stage but by adding this package HERE you include the lib into the stages.

**Review by davidmehren — APPROVED** [review]

**Comment by mrdrogdrog** [comment]
What about the token bar?

**Comment by mrdrogdrog** [comment]
Btw: the official remote cache is hosted by vercel. It seems like they do this for free. So we don't need to host the cache ourself. 🤷

**Comment by mrdrogdrog** [comment]
The test workflow hasn't been adjusted.

Remember to include the node version into the turbo dependencies

**Comment by mrdrogdrog** [comment]
How do we handle flakiness? You can force the rerun of tests.. Should we consider this somehow?

**Comment by ErikMichelson** [comment]
@davidmehren The required workflows need to be adjusted for allowing merges.

**Review by mrdrogdrog — APPROVED** [review]

**Review by DerMolly — APPROVED** [review]

**Review by ErikMichelson — COMMENTED** [review]
LGTM ✅

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+      - name: Login to GHCR
+        uses: docker/login-action@f4ef78c080cd8ba55a85445d5b36e214a81df20a # v2.1.0
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push backend
+        uses: docker/build-push-action@c56af957549030174b10d6867f20e78cfd7debc5 # v3.2.0
```
Comment (davidmehren) [inline .github/workflows/docker.yml:61]: This should probably use the same version as line 76?

File: .github/workflows/docker.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+jobs:
+  build-and-push:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+
+      - name: Generate Docker metadata for backend
+        id: meta-backend
+        uses: docker/metadata-action@57396166ad8aefe6098280995947635806a0e6ea # v4.1.1
```
Comment (davidmehren) [inline .github/workflows/docker.yml:27]: This should probably use the same version as line 38?

File: docs/content/dev/getting-started.md [hunk 4]
```diff
@@ -24,22 +24,21 @@ recommend caddy and the provided configuration.
...
-  echo "HD_AUTH_LOCAL_ENABLE_LOGIN=true" >> .env
-  echo "HD_AUTH_LOCAL_ENABLE_REGISTER=true" >> .env
-   ```
-11. Start the backend using `yarn start:dev`
-12. Go to  `hedgedoc/frontend` directory with `cd ../frontend`
-13. Start the frontend using `yarn run dev:with-local-backend`
-14. Go to `hedgedoc/dev-reverse-proxy` with `cd ../dev-reverse-proxy`
-15. Start Caddy using `./caddy run` (if you've downloaded the binary manually) or `caddy run` (if you've installed Caddy using a package manager)
-16. Use your browser to go to <http://localhost:8080>
+5. Goto `hedgedoc/backend` directory with `cd ../backend`
```
Comment (ErikMichelson) [inline docs/content/dev/getting-started.md:27]: Why `../backend`? Following the guide step-by-step, we're still in the repo root directory.
Furthermore, we first download caddy to `hedgedoc/dev-reverse-proxy` and then create the `hedgedoc` folder afterwards by the `git clone`. This won't work.

File: frontend/docker/Dockerfile [hunk 5]
```diff
@@ -2,8 +2,15 @@
 #
 # SPDX-License-Identifier: CC-BY-SA-4.0
 
+# BASE
+FROM docker.io/node:18-alpine@sha256:bc329c7332cffc30c2d4801e38df03cbfa8dcbae2a7a52a449db104794f168a3 AS base
+RUN apk add --no-cache tini
+ENTRYPOINT ["tini"]
```
Comment (ErikMichelson) [inline frontend/docker/Dockerfile:8]: The [official documentation](https://github.com/krallin/tini#using-tini) sets the entrypoint slightly different.

```suggestion
ENTRYPOINT ["tini", "--"]
```

File: package.json [hunk 6]
```diff
@@ -5,8 +5,15 @@
...
     "backend",
     "frontend",
-    "commons"
+    "commons",
+    "dev-reverse-proxy"
   ],
+  "scripts": {
+    "build": "turbo run build",
+    "lint": "turbo run lint",
+    "format": "turbo run format",
```
Comment (ErikMichelson) [inline package.json:14]: This errors:
```
 ERROR  run failed: task `format` not found in turbo `pipeline` in "turbo.json". Are you sure you added it?
```

**Review by ErikMichelson — COMMENTED** [review]
Seems good now

**Pull request merged**